llvm/lib/Target/X86/X86ISelDAGToDAG.cpp

```diff
@@ -1036,6 +1036,42 @@ void X86DAGToDAGISel::PreprocessISelDAG() {
 
       break;
     }
+    case ISD::LOAD: {
+      // If this is a XMM/YMM load of the same lower bits as another YMM/ZMM
+      // load, then just extract the lower subvector and avoid the second load.
+      auto *Ld = cast<LoadSDNode>(N);
+      MVT VT = N->getSimpleValueType(0);
+      if (ISD::isNormalLoad(Ld) && Ld->isSimple() &&
+          (VT.is128BitVector() || VT.is256BitVector())) {
+        SDValue Ptr = Ld->getBasePtr();
+        SDValue Chain = Ld->getChain();
+        for (SDNode *User : Ptr->uses()) {
+          auto *UserLd = dyn_cast<LoadSDNode>(N);
```

Comment:
Do you not need some check that there is no store between the two loads?

Reply:
We check that the chains match

Reply:
I.e could you have `load; store; load` all chained together?

Reply:
If the store changes the value for the second load, I think they will not have the same chain.

Reply:
store node always builds a new chain with the node.